packages/core/src/transport/httpRequest.spec.ts

```diff
@@ -19,62 +19,185 @@ describe('httpRequest', () => {
+      spyOn(navigator, 'sendBeacon').and.callFake(() => false)
+
+      request.send('{"foo":"bar1"}\n{"foo":"bar2"}', 10)
+
+      expect(requests.length).toEqual(1)
+      expect(requests[0].type).toBe('xhr')
+    })
 
-  it('should use xhr over sendBeacon when the bytes count is too high', () => {
-    request.send('{"foo":"bar1"}\n{"foo":"bar2"}', BATCH_BYTES_LIMIT)
+    it('should fallback to xhr when sendBeacon throws', () => {
+      if (!navigator.sendBeacon) {
+        pending('no sendBeacon support')
+      }
+      spyOn(navigator, 'sendBeacon').and.callFake(() => {
+        throw new TypeError()
+      })
 
-    expect(requests.length).toEqual(1)
-    expect(requests[0].type).toBe('xhr')
+      request.send('{"foo":"bar1"}\n{"foo":"bar2"}', 10)
+
+      expect(navigator.sendBeacon).toHaveBeenCalled()
+      expect(requests.length).toEqual(1)
+      expect(requests[0].type).toBe('xhr')
+    })
   })
 
-  it('should fallback to xhr when sendBeacon is not queued', () => {
-    if (!navigator.sendBeacon) {
-      pending('no sendBeacon support')
-    }
-    spyOn(navigator, 'sendBeacon').and.callFake(() => false)
+  describe('send (with fetch-keepalive FF)', () => {
+    beforeEach(() => {
+      updateExperimentalFeatures(['fetch-keepalive'])
+    })
+
+    afterEach(() => {
+      resetExperimentalFeatures()
+    })
 
-    request.send('{"foo":"bar1"}\n{"foo":"bar2"}', 10)
+    it('should use xhr when fetch keepalive is not available', () => {
+      const originalRequest = window.Request
+      ;(window.Request as any) = false
 
-    expect(requests.length).toEqual(1)
-    expect(requests[0].type).toBe('xhr')
+      request.send('{"foo":"bar1"}\n{"foo":"bar2"}', 10)
+
+      expect(requests.length).toEqual(1)
+      expect(requests[0].type).toBe('xhr')
+      expect(requests[0].url).toContain(ENDPOINT_URL)
+      expect(requests[0].body).toEqual('{"foo":"bar1"}\n{"foo":"bar2"}')
+      window.Request = originalRequest
+    })
+
+    it('should use fetch keepalive when the bytes count is correct', () => {
+      if (!('fetch' in window) || !('keepalive' in new window.Request(''))) {
+        pending('no fetch keepalive support')
+      }
+
+      request.send('{"foo":"bar1"}\n{"foo":"bar2"}', 10)
+
+      expect(requests.length).toEqual(1)
+      expect(requests[0].type).toBe('fetch')
+    })
+
+    it('should use xhr over fetch keepalive when the bytes count is too high', () => {
+      request.send('{"foo":"bar1"}\n{"foo":"bar2"}', BATCH_BYTES_LIMIT)
+
+      expect(requests.length).toEqual(1)
+      expect(requests[0].type).toBe('xhr')
+    })
+
+    it('should fallback to xhr when fetch keepalive is not queued', (done) => {
+      if (!('fetch' in window) || !('keepalive' in new window.Request(''))) {
+        pending('no fetch keepalive support')
+      }
+
+      let notQueuedFetch: Promise<never>
+      spyOn(window, 'fetch').and.callFake(() => {
+        notQueuedFetch = Promise.reject()
+        return notQueuedFetch
+      })
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏`jasmine.getEnv().allowRespy(true)` bothers me a bit. In the past, I got yield at by jasmine because some spies were not correctly cleaned up. By allowing "respy" globally, I think it opens the door to more mistakes and overly complex patterns.

What about something like this:

```ts
// in beforeEach
const { requests, mockNextRequest } = interceptRequests()

// in test case
mockNextRequest('fetch', () => Promise.reject())
```